subprojects/jvm-services/src/main/java/org/gradle/jvm/toolchain/internal/MavenToolchainsInstallationSupplier.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain.internal;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.provider.ProviderFactory;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import javax.inject.Inject;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class MavenToolchainsInstallationSupplier extends AutoDetectingInstallationSupplier {
+
+    private final Logger logger;
+    private final String defaultToolchainLocation;
+    private final String parseExpression;
+    private final XPathFactory xPathFactory;
+    private final DocumentBuilderFactory documentBuilderFactory;
+
+    @Inject
+    public MavenToolchainsInstallationSupplier(ProviderFactory factory) {
+        super(factory);
+        logger = Logging.getLogger(MavenToolchainsInstallationSupplier.class);
+        defaultToolchainLocation = new File(System.getProperty("user.home"), ".m2/toolchains.xml").getAbsolutePath();
```

Comment:
How do you feel about allowing for non-default locations too via a Gradle property like `org.gradle.java.installations.maven-toolchains-file`?

Reply:
I thought about that but opted out of implementing this because I wanted a minimal working implementation first. Maven itself does support non-default locations (advises against them though) so supporting that here as well does make sense. I'll take a look at adding support for the proposed property later today or tomorrow.

Reply:
Thanks, no rush. I think this would be a small enough change not worth leaving out.